### DISCUSSION TITLE
Conversion to const char*

### DISCUSSION THREAD

**Discussion by Kolhun** [nlohmann/json#4136] [Q&A] [answered]
At startup, the program crashes, I assume that the problem is in the convert method (if you do as in the first argument, the conversion first into a string and then into characters, then everything is OK) How can this be corrected? Code below 

(Yes, the setValue method takes const char*)

///
const char* convert(std::string content)
{
    const char* defined = content.c_str();
    return defined;
}

![image](https://github.com/nlohmann/json/assets/36815965/05a4fc56-e6a2-4b89-a7a5-a21178e29aa8)
![image](https://github.com/nlohmann/json/assets/36815965/f470f066-c83e-4276-9208-875545a61f9c)

**Comment by gregmarr** [comment]
This has nothing to do with the library.  Your convert function is returning a pointer from a temporary std::string (it's passed by value instead of reference), so it's invalid as soon as the function returns.

**Reply by Kolhun** [reply]
Oh! Thanks

**Reply by Kolhun** [reply]
I try and it doesn't work, maybe I'm not doing it right?
![image](https://github.com/nlohmann/json/assets/36815965/ea55ee79-cbe4-4307-885d-7a2bc363fa9a)

**Reply by gregmarr** [reply]
If that is all that your `convert` function is doing, then you don't even need the function, just call `.c_str()` directly.

**Reply by Kolhun** [reply]
Yes, that's all it does, and no, it doesn't work that way either

![image](https://github.com/nlohmann/json/assets/36815965/2d5cfa30-4e66-47a0-ab9d-af2f8596ef4d)

The point is that the section of code associated with defId is working properly, but the method is not

![image](https://github.com/nlohmann/json/assets/36815965/0bab7e3f-e73d-4445-978e-0a99685da6ce)

**Reply by gregmarr** [reply, marked as answer]
That's because `dataResult["id"]` isn't a `std::string`, it's a `json`.  You need to convert it to a `std::string` as you did with `_id`.

    dataResult["id"].get<std::string>().c_str()

or to be safer if you don't know for sure that there is an `"id"` child in `dataResult`

    dataResult.value<std::string>("id", "some default value").c_str()

Note that this is STILL a temporary, and it will be destroyed as soon as the statement finishes, so `ini.SetValue` needs to copy the string, but at least now it survives the call to `ini.SetValue()`.  This applies to both `const char *` parameters to `ini.SetValue()`, as `defId` dies when `_id` is destroyed.

**Reply by Kolhun** [reply]
That's how, and how can this be wrapped into a method? I tried to get into it, it didn't work out very well :D

![image](https://github.com/nlohmann/json/assets/36815965/c1b70597-fcce-4dc1-ba67-bcfc4bb340b4)

**Reply by gregmarr** [reply]
You can't do that, as then you lose the magic lifetime extension.

You can do it with `get_ref` instead of `get`, so then the `const char *` lives as long as the `json`, which is better.  https://www.godbolt.org/z/hWsYqceMK

**Reply by Kolhun** [reply]
Well, thank you, you helped me a lot)))
